ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py

```diff
@@ -101,127 +111,280 @@ def _evaluate(self, datasets, **kwargs):
+            # as predicted Application
+            pdfAppPred = knlMsr(Y[0, :])
+        else:
+            # If not, KDE of Experiment and Application
+            knl = stats.gaussian_kde(vals)
+            # Joint PDF of Experiment and Application
+            Z = np.reshape(knl(psts).T, X.shape)
+            # yAppPred by integrating p(yexp, yapp)p(ymsr) over [yexp.min(), yexp.max()]
+            pdfAppPred = np.dot(Z, pdfMsr.reshape(pdfMsr.shape[0], 1))
+
+        # Normalized PDF of predicted application
+        pdfAppPredNorm = pdfAppPred.flatten()/pdfAppPred.sum()/np.diff(Y[0, :])[0]
+
+        # Calculate Expectation (average value) of predicted application
+        # by integrating xf(x), where f(x) is PDF of x
+        predMean = 0.0
+        for i in range(len(Y[0, :])):
+            predMean += Y[0, i]*pdfAppPredNorm[i]*(Y[0, 1]-Y[0, 0])
+
+        # Calculate Variance of predicted application
+        # by integrating (x-mu_x)^2f(x), where f(x) is PDF of x
+        predVar = 0.0
+        for i in range(len(Y[0, :])):
+            predVar += (Y[0, i]-predMean)**2.0 * pdfAppPredNorm[i]*(Y[0, 1]-Y[0, 0])
+
+        # Predicted standard deviation is square root of variance
+        predStd = np.sqrt(predVar)
+        # Prior standard deviation is the sample standard deviation
+        # Consider probability weights in the future
+        priStd = np.std(yAppStd)
+        # Uncertainty reduction fraction is 1.0-sigma_pred/sigma_pri
+        outputArray.append(1.0-predStd/priStd)
+      return outputArray
+
+    # functions for T-Dep PCM
+    def KNN(exp, app, expM, k=5, Metric='minkowski'):
+      # check the dimension of exp matrix, \
+      # if it only has one feature, enasure to make it as a 2D matrix instead of a 1D vector
+      if expM.ndim==1:
+          expM = expM.reshape(1, expM.shape[0])
+      # Scale exp data with its mean (mu) and standard deviation (std).  [expS = (exp - mu)/std]
+      # Scale expM data with by same mu and std. [expMS = (expM - mu)/std]
+      scaler = StandardScaler()
+      expS = scaler.fit_transform(exp)
+      expMS = scaler.transform(expM)
+      # Fit KNN regression model
+      model = KNeighborsRegressor(n_neighbors=k, metric=Metric)
+      model.fit(expS, app)
+      # Predict on measurement set
+      appPred = model.predict(expMS)
+
+      return appPred
+
+    def FindRank(ref, x, xM):
+      x = (x.T/ref).T
+      error = (np.std(xM, axis=0)/ np.mean(xM, axis=0))[0] #msr error
+      #scaling by reference
+      u, s, v_T = np.linalg.svd(x)
+      # Find the maximum of the residual under different rank
+      maxError = np.zeros(np.mean(xM, axis=0).shape[0])
+      for k in range(1,maxError.shape[0]+1):
+        res = x - u[:,:k]@np.diag(s)[:k,:k]@v_T[:k,:]#residual under different rank
+        maxError[k-1] = np.max(abs(res)) # maximum of the residual
+      xMAX=min(np.argwhere(maxError<error/10)) #rank corresponding to msr error
```

Comment:
The prediction of pcmTdep is sensitive to the rank, which determines the dimension of U_exp, U_app subspaces for KNN regression. We will lose information under low rank, or overfit data with noise under high rank. 1/10 is decided here to ensure a rank corresponding to around 0.1% reconstruction error, because the error from SETH measurement data is around 2% which is too big for deciding on reconstruction error. I think I could make a new node in input XML file to make this error user-defined later, instead of calculated from measurement data.

Reply:
Could you add a new xml node to control this?